File: backend/src/domain/repository/IFundRepository.ts
import { Fund } from '../entities/Fund';
import { FundReceivingInfo } from '../entities/FundReceivingInfo';
import { FundDonation } from '../entities/FundDonation';
import { FundExpense } from '../entities/FundExpense';

/**
 * Repository interface for fund and donation management operations
 */
export interface IFundRepository {
  // Fund Management
  createFund(fundData: Partial<Fund>): Promise<Fund>;
  updateFund(fundId: number, fundData: Partial<Fund>): Promise<Fund>;
  deleteFund(fundId: number): Promise<boolean>;
  findFundById(fundId: number): Promise<Fund | null>;
  findFundsByOrganization(
    organizationId: number,
    page: number,
    limit: number,
    filters?: { status?: string },
  ): Promise<{ funds: Fund[]; total: number }>;
  updateFundAmount(fundId: number, amount: number): Promise<Fund>;
  updateFundStatus(fundId: number, status: string): Promise<Fund>;

  // Fund Receiving Info Management
  createReceivingInfo(
    receivingInfoData: Partial<FundReceivingInfo>,
  ): Promise<FundReceivingInfo>;
  updateReceivingInfo(
    receivingInfoId: number,
    receivingInfoData: Partial<FundReceivingInfo>,
  ): Promise<FundReceivingInfo>;
  deleteReceivingInfo(receivingInfoId: number): Promise<boolean>;
  findReceivingInfosByFund(fundId: number): Promise<FundReceivingInfo[]>;
  findActiveReceivingInfosByFund(fundId: number): Promise<FundReceivingInfo[]>;
  toggleReceivingInfoStatus(
    receivingInfoId: number,
    isActive: boolean,
  ): Promise<FundReceivingInfo>;

  // Donation Management
  createDonation(donationData: Partial<FundDonation>): Promise<FundDonation>;
  updateDonation(
    donationId: number,
    donationData: Partial<FundDonation>,
  ): Promise<FundDonation>;
  findDonationById(donationId: number): Promise<FundDonation | null>;
  findDonationsByFund(
    fundId: number,
    page: number,
    limit: number,
    filters?: { status?: string },
  ): Promise<{ donations: FundDonation[]; total: number }>;
  findDonationsByDonor(
    donorMemberId: number,
    page: number,
    limit: number,
  ): Promise<{ donations: FundDonation[]; total: number }>;
  approveDonation(donationId: number): Promise<FundDonation>;
  rejectDonation(donationId: number): Promise<FundDonation>;

  // Expense Management
  createExpense(expenseData: Partial<FundExpense>): Promise<FundExpense>;
  updateExpense(
    expenseId: number,
    expenseData: Partial<FundExpense>,
  ): Promise<FundExpense>;
  deleteExpense(expenseId: number): Promise<boolean>;
  findExpenseById(expenseId: number): Promise<FundExpense | null>;
  findExpensesByFund(
    fundId: number,
    page: number,
    limit: number,
  ): Promise<{ expenses: FundExpense[]; total: number }>;

  // Statistics & Reports
  getFundStatistics(fundId: number): Promise<{
    totalDonations: number;
    approvedDonations: number;
    pendingDonations: number;
    totalExpenses: number;
    currentBalance: number;
    donorCount: number;
  }>;

  getTotalDonationsByFund(fundId: number): Promise<number>;
  getTotalExpensesByFund(fundId: number): Promise<number>;
}

File: backend_java/src/main/java/com/service/backend/domain/repository/IFundRepository.java
package com.service.backend.domain.repository;

import com.service.backend.domain.entity.Fund;
import com.service.backend.domain.entity.FundReceivingInfo;
import com.service.backend.domain.entity.FundDonation;
import com.service.backend.domain.entity.FundExpense;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.math.BigDecimal;
import java.util.Map;

/**
 * Repository interface for fund and donation management operations
 */
public interface IFundRepository {

    // Fund Management
    Mono<Fund> createFund(Fund fundData);
    Mono<Fund> updateFund(Long fundId, Fund fundData);
    Mono<Boolean> deleteFund(Long fundId);
    Mono<Fund> findFundById(Long fundId);
    Mono<Map<String, Object>> findFundsByOrganization(Long organizationId, int page, int limit, Map<String, Object> filters);
    Mono<Fund> updateFundAmount(Long fundId, BigDecimal amount);
    Mono<Fund> updateFundStatus(Long fundId, String status);

    // Fund Receiving Info Management
    Mono<FundReceivingInfo> createReceivingInfo(FundReceivingInfo receivingInfoData);
    Mono<FundReceivingInfo> updateReceivingInfo(Long receivingInfoId, FundReceivingInfo receivingInfoData);
    Mono<Boolean> deleteReceivingInfo(Long receivingInfoId);
    Flux<FundReceivingInfo> findReceivingInfosByFund(Long fundId);
    Flux<FundReceivingInfo> findActiveReceivingInfosByFund(Long fundId);
    Mono<FundReceivingInfo> toggleReceivingInfoStatus(Long receivingInfoId, Boolean isActive);

    // Donation Management
    Mono<FundDonation> createDonation(FundDonation donationData);
    Mono<FundDonation> updateDonation(Long donationId, FundDonation donationData);
    Mono<FundDonation> findDonationById(Long donationId);
    Mono<Map<String, Object>> findDonationsByFund(Long fundId, int page, int limit, Map<String, Object> filters);
    Mono<Map<String, Object>> findDonationsByDonor(Long donorMemberId, int page, int limit);
    Mono<FundDonation> approveDonation(Long donationId);
    Mono<FundDonation> rejectDonation(Long donationId);

    // Expense Management
    Mono<FundExpense> createExpense(FundExpense expenseData);
    Mono<FundExpense> updateExpense(Long expenseId, FundExpense expenseData);
    Mono<Boolean> deleteExpense(Long expenseId);
    Mono<FundExpense> findExpenseById(Long expenseId);
    Mono<Map<String, Object>> findExpensesByFund(Long fundId, int page, int limit);

    // Statistics & Reports
    Mono<Map<String, Object>> getFundStatistics(Long fundId);
    Mono<BigDecimal> getTotalDonationsByFund(Long fundId);
    Mono<BigDecimal> getTotalExpensesByFund(Long fundId);
}
